text/0000-inline-asm.md

```diff
@@ -0,0 +1,892 @@
+    : "a"(4), "c"(0) // inputs
+    : "edx" // clobbers
+);
+printf("L1 Cache: %i\n", ((ebx >> 22) + 1)
+    * (((ebx >> 12) & 0x3ff) + 1)
+    * ((ebx & 0xfff) + 1)
+    * (ecx + 1));
+```
+
+## Limited set of operand types
+
+The proposed set of operand types is much smaller than that which is available through GCC-style inline assembly. In particular, the proposed syntax does not include any form of memory operands and is missing many register classes.
+
+We chose to keep operand constraints as simple as possible, and in particular memory operands introduce a lot of complexity since different instruction support different addressing modes. At the same time, the exact rules for memory operands are not very well known (you are only allowed to access the data directly pointed to by the constraint) and are often gotten wrong.
+
+If we discover that there is a demand for a new register class or special operand type, we can always add it later.
+
+## Difficulty of support
+
+Inline assembly is a difficult feature to implement in a compiler backend. While LLVM does support it, this may not be the case for alternative backends such as [Cranelift][cranelift] (see [this issue][cranelift-asm]).
+
+However it is possible to implement support for inline assembly without support from the compiler backend by using an external assembler instead. Take the following (AArch64) asm block as an example:
+
+```rust
+unsafe fn foo(mut a: i32, b: i32) -> (i32, i32)
+{
+    let c;
+    asm!("<some asm code>", inout(reg) a, in("x0") b, out("x20") c);
+    (a, c)
+}
+```
+
+This could be expanded to an external asm file with the following contents:
+
+```
+# Function prefix directives
+.section ".text.foo_inline_asm"
+.globl foo_inline_asm
+.p2align 2
+.type foo_inline_asm, @function
+foo_inline_asm:
+
+// If necessary, save callee-saved registers to the stack here.
+str x20, [sp, #-16]!
+
+// Move the pointer to the argument out of the way since x0 is used.
+mov x1, x0
+
+// Load inputs values
+ldr w2, [x1, #0]
+ldr w0, [x1, #4]
+
+<some asm code>
+
+// Store output values
+str w2, [x1, #0]
+str w20, [x1, #8]
+
+// If necessary, restore callee-saved registers here.
+ldr x20, [sp], #16
+
+ret
+
+# Function suffix directives
+.size foo_inline_asm, . - foo_inline_asm
+```
+
+And the following Rust code:
+
+```rust
+unsafe fn foo(mut a: i32, b: i32) -> (i32, i32)
+{
+    let c;
+    {
+        #[repr(C)]
+        struct foo_inline_asm_args {
+            a: i32,
+            b: i32,
+            c: i32,
+        }
+        extern "C" {
+            fn foo_inline_asm(args: *mut foo_inline_asm_args);
+        }
+        let mut args = foo_inline_asm_args {
+            a: a,
+            b: b,
+            c: mem::uninitialized(),
+        };
+        foo_inline_asm(&mut args);
+        a = args.a;
+        c = args.c;
+    }
+    (a, c)
+}
+```
+
+[cranelift]: https://cranelift.readthedocs.io/
+[cranelift-asm]: https://github.com/bytecodealliance/cranelift/issues/444
+
```

Comment:
```suggestion
If `rustc` does not support inline assembly for a particular target and backend, trying to use the `asm!` macro on that particular backend and target fails to compile.
```

For example, consider we actually do this, and add support for using `asm!` with Cranelift and `x86_64`. What happens if Cranelift implements `aarch64` support, and somebody tries to compile a program for that target using that backend ? 

Should we forbid trying to do that? That is, should be forbid adding rustc support for new targets until `rustc` supports inline assembly for them? Or should we allow that as long as the programs being compiled do not use the `asm!` macro ? 

I think that blocking target/backend support on having inline assembly worked out imposes a too high barrier of entry for new targets. E.g., this RFC adds x86, arm, and risc-v support, but there are people using PowerPC and others on stable, which, according to this RFC, wouldn't initially gain any `asm!` macro support, yet we can't break that code.

Reply:
My reading of that is simply that it is a compile error, not that the target can't be added at all.

Reply:
Where did you read that ?

Reply:
It says using `asm!` when building for a target without `asm` support is an error:
* So there must _be_ targets without `asm` support that otherwise can do builds.
* And building for them is fine as long as you don't use `asm`

Reply:
Where does the RFC say that? (I've searched for that info a couple of times and can't find it, can you provide a link to the actual line?).

Reply:
I think you are in violent agreement.  If I'm interpreting GitHub's interface properly @gnzlbg is suggesting to add the highlighted line since it isn't yet part of the RPC. Unfortunately, for other readers it looks like the comment is attached to an existing sentence from the proposal.

Reply:
OH.

I thought the 2nd quote block was already in the RFC! Stupid GitHub interface.

Reply:
Duh, I failed at this too. 

FWIW, @lokathor is much better at reading than me, and it is often the case that I miss something that's literally written "as is" in the text, and @lokathor just points me to it 😆 

So I actually was searching the whole RFC for this thinking "not again..." 😆